From: Jan Höppner <hoeppner@linux.vnet.ibm.com>
Subject: s390/dasd: Simplify code in format logic
Patch-mainline: v4.6-rc1
Git-commit: 8542885bf5338bfccf779868a5d143620e794ff9
References: bsc#1025482, FATE#321527, LTC#146542

Summary:     dasd: Add new ioctl BIODASDCHECKFMT
Description: Implement new DASD IOCTL BIODASDCHECKFMT to check a range of
             tracks on a DASD volume for correct formatting. The following
             characteristics are checked:
             - Block size
             - ECKD key length
             - ECKD record ID
             - Number of records per track


Upstream-Description:

             s390/dasd: Simplify code in format logic

             Currently, dasd_format is calling the format logic of a DASD discipline
             with PAV enabled. If that fails with an error code of -EAGAIN the value
             of retries is decremented and the discipline function is called with PAV
             turned off.

             The loop is supposed to try this up to 255 times until success. However,
             -EAGAIN can only occur once here and therefore the loop will never reach
             the 255 retries.

             So, replace the unnecessarily complicated loop logic and simply try again
             without PAV enabled in case of an -EAGAIN error.

             Signed-off-by: Jan Höppner <hoeppner@linux.vnet.ibm.com>
             Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Jan Höppner <hoeppner@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.com>
---
 drivers/s390/block/dasd_ioctl.c |   33 ++++++---------------------------
 1 file changed, 6 insertions(+), 27 deletions(-)

--- a/drivers/s390/block/dasd_ioctl.c
+++ b/drivers/s390/block/dasd_ioctl.c
@@ -203,9 +203,7 @@ static int
 dasd_format(struct dasd_block *block, struct format_data_t *fdata)
 {
 	struct dasd_device *base;
-	int enable_pav = 1;
-	int rc, retries;
-	int start, stop;
+	int rc;
 
 	base = block->base;
 	if (base->discipline->format_device == NULL)
@@ -233,30 +231,11 @@ dasd_format(struct dasd_block *block, st
 		bdput(bdev);
 	}
 
-	retries = 255;
-	/* backup start- and endtrack for retries */
-	start = fdata->start_unit;
-	stop = fdata->stop_unit;
-	do {
-		rc = base->discipline->format_device(base, fdata, enable_pav);
-		if (rc) {
-			if (rc == -EAGAIN) {
-				retries--;
-				/* disable PAV in case of errors */
-				enable_pav = 0;
-				fdata->start_unit = start;
-				fdata->stop_unit = stop;
-			} else
-				return rc;
-		} else
-			/* success */
-			break;
-	} while (retries);
-
-	if (!retries)
-		return -EIO;
-	else
-		return 0;
+	rc = base->discipline->format_device(base, fdata, 1);
+	if (rc == -EAGAIN)
+		rc = base->discipline->format_device(base, fdata, 0);
+
+	return rc;
 }
 
 /*
